### ISSUE TITLE
Windows failures summary

### ISSUE THREAD

**Issue by ugexe** [perl6/roast#504]
tl;dr -- Majority of failures seem related to mode and permissions related to filesystem IO. There are a couple interesting failures though.

---

```
t\spec\S02-magicals\KERNEL.rakudo.moar                          (Wstat: 512 Tests: 42 Failed: 2)
  Failed tests:  39-40
  Non-zero exit status: 2
```

Tested signals not available on windows

https://github.com/perl6/roast/blob/2fc3fb676284d3dbd058a2e3c7bc6830c8fb65d2/S02-magicals/KERNEL.t#L70-L71

---

```
t\spec\S02-types\num.rakudo.moar                                (Wstat: 256 Tests: 108 Failed: 1)
  Failed test:  104
  Non-zero exit status: 1
```

`cmp-ok Num(0.777777777777777777777), '==', Num(0.7777777777777777777771), 'Rat->Num conversion is monotonic';`

https://github.com/perl6/roast/blob/2fc3fb676284d3dbd058a2e3c7bc6830c8fb65d2/S02-types/num.t#L853

---

```
t\spec\S04-declarations\constant-6.d.t                          (Wstat: 2304 Tests: 0 Failed: 0)
  Non-zero exit status: 9
  Parse errors: No plan found in TAP output
```

Test exits almost immediately with no output. No idea why this only happens only windows unless these 3 lines are commented out:

  * https://github.com/perl6/roast/blob/2fc3fb676284d3dbd058a2e3c7bc6830c8fb65d2/S04-declarations/constant-6.d.t#L518
  * https://github.com/perl6/roast/blob/2fc3fb676284d3dbd058a2e3c7bc6830c8fb65d2/S04-declarations/constant-6.d.t#L581
  * https://github.com/perl6/roast/blob/2fc3fb676284d3dbd058a2e3c7bc6830c8fb65d2/S04-declarations/constant-6.d.t#L644

---

```
t\spec\S06-other\main-usage.rakudo.moar (Wstat: 256 Tests: 43 Failed: 1)
  Failed test:  17
  Non-zero exit status: 1
Files=1, Tests=43, 14 wallclock secs ( 0.08 usr +  0.00 sys =  0.08 CPU)
Result: FAIL
```

`:$y` appears to be an empty string inside MAIN when passed `"-y=\n"`

https://github.com/perl6/roast/blob/2fc3fb676284d3dbd058a2e3c7bc6830c8fb65d2/S06-other/main-usage.t#L68-L69

---

```
t\spec\S16-filehandles\filetest.rakudo.moar                     (Wstat: 5632 Tests: 128 Failed: 22)
  Failed tests:  57-64, 81-82, 85-86, 89-90, 93-94, 105-108
                113-114
  Non-zero exit status: 22
```

Mode/permissions issues?

---

```
t\spec\S16-filehandles\mode.t                                   (Wstat: 768 Tests: 12 Failed: 3)
  Failed tests:  3, 6, 12
  Non-zero exit status: 3
```

Certain mode settings don't work on windows

---

```
t\spec\S32-encoding\decoder.t                                   (Wstat: 256 Tests: 2 Failed: 1)
  Failed test:  2
  Non-zero exit status: 1
```

No idea. process exitcode is 1, and stdout and stderr are both empty

https://github.com/perl6/roast/blob/2fc3fb676284d3dbd058a2e3c7bc6830c8fb65d2/S32-encoding/decoder.t#L33

---

```
t\spec\S32-io\chdir.t                                           (Wstat: 5376 Tests: 82 Failed: 21)
  Failed tests:  30-33, 43, 45, 47, 49, 54-55, 59, 61-65
                67, 71, 75, 78-79
  Non-zero exit status: 21
```

Mode/permissions issues?

---

```
t\spec\S32-io\file-tests.t                                      (Wstat: 1280 Tests: 16 Failed: 5)
  Failed tests:  6, 8-10, 15
  Non-zero exit status: 5
```

Mode/permissions issues?

---

```
t\spec\S32-io\indir.t                                           (Wstat: 5376 Tests: 77 Failed: 21)
  Failed tests:  22-25, 35, 37, 39, 41, 46-47, 51, 53-57
                59, 63, 67, 70-71
  Non-zero exit status: 21
```

Mode/permissions issues?

---

```
t\spec\S32-io\io-cathandle.rakudo.moar                          (Wstat: 256 Tests: 30 Failed: 1)
  Failed test:  24
  Non-zero exit status: 1
```

Somehow it ends up prefixing an absolute url with the $*CWD ( C:\foo\bar\baz\C:\foo\bar\baz\bip\mow )

---

```
t\spec\S32-io\io-spec-win.t                                     (Wstat: 512 Tests: 213 Failed: 2)
  Failed tests:  183-184
  Non-zero exit status: 2
```

Both failing tests are for abs2rel when using a UNC(?) path
```
    # Failed test 'abs2rel: $("\\\\a/t1/t2/t4", "/t2/t3") -> '\\a\t1\t2\t4''
    # at t\spec\S32-io\io-spec-win.t line 266
    # expected: '\\a\t1\t2\t4'
    #      got: '..\..\a\t1\t2\t4'
```
https://github.com/perl6/roast/blob/2fc3fb676284d3dbd058a2e3c7bc6830c8fb65d2/S32-io/io-spec-win.t#L249-L250

---

```
t\spec\S32-io\open.t                                            (Wstat: 256 Tests: 16 Failed: 1)
  Failed test:  15
  Non-zero exit status: 1
  Parse errors: Bad plan.  You planned 17 tests but ran 16.
```

On a windows machine i'm testing with this test fails at 933 .Caused by the %s/fname used in MVM_exception_throw_adhoc_free

```
> C:\Users\ugexe\.rakudobrew\moar-blead-master>perl6 -e "('a' x 975).IO.open"
    MoarVM panic: Memory allocation failed; could not allocate 18446744073709551612 bytes
```

https://github.com/perl6/roast/blob/2fc3fb676284d3dbd058a2e3c7bc6830c8fb65d2/S32-io/open.t#L241-L243

---

```
t\spec\S32-io\pipe.t                                            (Wstat: 768 Tests: 20 Failed: 3)
  Failed tests:  12-13, 19
  Non-zero exit status: 3
```

Chaining stdin to stdout `run(..., :in($proc.out))` doesn't work

https://github.com/perl6/roast/blob/2fc3fb676284d3dbd058a2e3c7bc6830c8fb65d2/S32-io/pipe.t#L46-L53

---

```
t\spec\S17-procasync\bind-handles.t                             (Wstat: 65280 Tests: 15 Failed: 0)
  Non-zero exit status: 255
  Parse errors: Bad plan.  You planned 17 tests but ran 15.
```

Same as t\spec\S32-io\pipe.t issue?

```
    A react block:
      in block <unit> at t\spec\S17-procasync\bind-handles.t line 117

    Died because of the exception:
        bad file descriptor
          in block <unit> at t\spec\S17-procasync\bind-handles.t line 117
```

https://github.com/perl6/roast/blob/2fc3fb676284d3dbd058a2e3c7bc6830c8fb65d2/S17-procasync/bind-handles.t#L113-L141

---

**Comment by dumarchie** [comment]
This is an update for Rakudo version 2019.11-795-g65ef7db19 built on MoarVM version 2020.01.1-47-gdd20883c3 running on Windows 10:

In three consecutive runs of `gmake spectest` the following tests failed consistently:

```
t\spec\S02-magicals\KERNEL.rakudo.moar                          (Wstat: 768 Tests: 42 Failed: 3)
  Failed tests:  24, 39-40
  Non-zero exit status: 3
t\spec\S16-filehandles\filetest.rakudo.moar                     (Wstat: 5632 Tests: 128 Failed: 22)
  Failed tests:  57-64, 81-82, 85-86, 89-90, 93-94, 105-108
                113-114
  Non-zero exit status: 22
t\spec\S16-filehandles\mode.t                                   (Wstat: 768 Tests: 12 Failed: 3)
  Failed tests:  3, 6, 12
  Non-zero exit status: 3
t\spec\S17-procasync\basic.rakudo.moar                          (Wstat: 256 Tests: 46 Failed: 1)
  Failed test:  38
  Non-zero exit status: 1
t\spec\S17-procasync\bind-handles.t                             (Wstat: 65280 Tests: 15 Failed: 0)
  Non-zero exit status: 255
  Parse errors: Bad plan.  You planned 17 tests but ran 15.
t\spec\S17-scheduler\every.t                                    (Wstat: 256 Tests: 25 Failed: 1)
  Failed test:  13
  Non-zero exit status: 1
t\spec\S19-command-line-options\02-dash-n.rakudo.moar           (Wstat: 256 Tests: 6 Failed: 1)
  Failed test:  3
  Non-zero exit status: 1
t\spec\S22-package-format\local.t                               (Wstat: 256 Tests: 19 Failed: 1)
  Failed test:  7
  Non-zero exit status: 1
t\spec\S26-documentation\block-leading-user-format.t            (Wstat: 512 Tests: 2 Failed: 2)
  Failed tests:  1-2
  Non-zero exit status: 2
t\spec\S32-io\chdir.t                                           (Wstat: 5376 Tests: 82 Failed: 21)
  Failed tests:  30-33, 43, 45, 47, 49, 54-55, 59, 61-65
                67, 71, 75, 78-79
  Non-zero exit status: 21
t\spec\S32-io\file-tests.t                                      (Wstat: 1280 Tests: 16 Failed: 5)
  Failed tests:  6, 8-10, 15
  Non-zero exit status: 5
t\spec\S32-io\indir.t                                           (Wstat: 5376 Tests: 77 Failed: 21)
  Failed tests:  22-25, 35, 37, 39, 41, 46-47, 51, 53-57
                59, 63, 67, 70-71
  Non-zero exit status: 21
t\spec\S32-io\io-cathandle.rakudo.moar                          (Wstat: 256 Tests: 30 Failed: 1)
  Failed test:  24
  Non-zero exit status: 1
t\spec\S32-io\io-path.t                                         (Wstat: 256 Tests: 39 Failed: 1)
  Failed test:  27
  Non-zero exit status: 1
t\spec\S32-io\io-spec-win.t                                     (Wstat: 512 Tests: 213 Failed: 2)
  Failed tests:  183-184
  Non-zero exit status: 2
t\spec\S32-io\open.t                                            (Wstat: 256 Tests: 16 Failed: 1)
  Failed test:  15
  Non-zero exit status: 1
  Parse errors: Bad plan.  You planned 17 tests but ran 16.
t\spec\S32-io\pipe.t                                            (Wstat: 512 Tests: 15 Failed: 2)
  Failed tests:  7-8
  Non-zero exit status: 2
```

The following tests failed occasionally:

```
t\spec\S02-names\pseudo-6c.rakudo.moar                          (Wstat: 256 Tests: 0 Failed: 0)
  Non-zero exit status: 1
  Parse errors: No plan found in TAP output
t\spec\S17-supply\zip-latest.t                                  (Wstat: 256 Tests: 0 Failed: 0)
  Non-zero exit status: 1
  Parse errors: No plan found in TAP output
t\spec\S17-supply\min.t                                         (Wstat: 256 Tests: 0 Failed: 0)
  Non-zero exit status: 1
  Parse errors: No plan found in TAP output
t\spec\S17-supply\schedule-on.t                                 (Wstat: 256 Tests: 0 Failed: 0)
  Non-zero exit status: 1
  Parse errors: No plan found in TAP output
t\spec\S32-io\lock.t                                            (Wstat: 256 Tests: 0 Failed: 0)
  Non-zero exit status: 1
  Parse errors: No plan found in TAP output
```

In earlier test runs `gmake spectest` would occasionally hang on one of those t\spec\S17-supply\ tests.
